### PR TITLE
feat(tests): add tests to EpiGraphHub webserver

### LINKED ISSUE #179: Add some selenium tests
it would be nice to have some simple selenium tests just to run some smoke tests

### PR DESCRIPTION
**Pull request by luabida**
Tests include:
- access epigraphhub.org
- click log in button
- input credentials 'guest:guest'
- click sign in button

fix #179

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -91,6 +91,16 @@ jobs:
         use-mamba: true
         miniforge-variant: Mambaforge
 
+    - uses: nanasess/setup-chromedriver@v1
+      - run: |
```
Comment (xmnlab) [inline .github/workflows/main.yaml:95]: here "uses" and "run" are 2 steps .. so move "run" to 1 indentation behind

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -91,6 +91,16 @@ jobs:
...
+          export DISPLAY=:99
+          chromedriver --url-base=/wd/hub &
+          sudo Xvfb -ac :99 -screen 0 1280x1024x24 > /dev/null 2>&1 & # optional
+
+    - name: run epigraphhub tests
+      - run: |
```
Comment (xmnlab) [inline .github/workflows/main.yaml:101]: here "run" is not a new step .. is just part of the "run epigraphhub test" step
so you don´t need a "-" here

File: tests/test_webapp.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import unittest
+import os
+
+CHROMEDRIVER = os.getenv('CHROMEDRIVER')
+
+class TestEpiGraphHub(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.driver = webdriver.Chrome(CHROMEDRIVER)
+        cls.driver.get("https://epigraphhub.org/")
```
Comment (xmnlab) [inline tests/test_webapp.py:12]: I guess here it needs to be http://localhost:8088

**Review by xmnlab — CHANGES_REQUESTED** [review]
Thanks for working on that!
In general it looks fine, just added one comment here

File: tests/test_webapp.py (hunk 3 above)
Comment (luabida) [reply tests/test_webapp.py:12]: I'm aware, and the step should be below the containers initialization. But the chromedriver nor chrome-setup seem to be working on CI, kinda lost at this point

**Comment by xmnlab** [comment]
I will provide some examples in a bit

**Comment by fccoelho** [comment]
Thanks for pushing this @xmnlab, this will be very important for the maintenability of the egh services

**Comment by luabida** [comment]
> I will provide some examples in a bit

@xmnlab t he last solution I have in mind is instead of using the chromedriver action, install the chrome with wget and run it as a step, any thoughts about trying this approach?

**Comment by xmnlab** [comment]
maybe you can find something here: https://github.com/osl-pocs/poc-python-selenium/

**Comment by luabida** [comment]
@xmnlab it seems like the postgres container is not initializing properly, dor you have any idea why the postgres volumes of the compose-dev are not working correctly?

**Comment by xmnlab** [comment]
I don't think that it is a volume problem the volume is not pointing to a specific path in the host side: https://github.com/thegraphnetwork/EpiGraphHub/blob/main/containers/compose-dev.yaml#L18

but if you want to test, you can remove that line and check how it would be performed by the CI

**Comment by luabida** [comment]
@xmnlab the file that has the line `GRANT USAGE ON SCHEMA switzerland TO dev_epigraph;` is one the volumes below that, more specifically the  `./postgresql/sql/dev/02-epigraphhub.sql`

**Comment by xmnlab** [comment]
Got it .. have you tried that locally as well?
Just to check where is the problem .. maybe a new version of the image changed the path etc

**Comment by luabida** [comment]
@xmnlab  Yes, the same error occurs locally

**Comment by xmnlab** [comment]
> @xmnlab Yes, the same error occurs locally

can you share what you have in the logs?
maybe it would give us some tips

PS: btw, in the worst case we could have it as a COPY statement in the dockerfile

**Comment by luabida** [comment]
> > @xmnlab can you share what you have in the logs? maybe it would give us some tips

sure, the logs are very similar
<details>
[2023-04-06, 17:47:18 UTC] {{__init__.py:117}} DEBUG - Preparing lineage inlets and outlets
[2023-04-06, 17:47:18 UTC] {{__init__.py:155}} DEBUG - inlets: [], outlets: []
[2023-04-06, 17:47:18 UTC] {{taskinstance.py:1775}} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.InsufficientPrivilege: permission denied for schema switzerland
LINE 1: select * from switzerland.foph_casesvaccpersons_d;
                      ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/airflow/operators/python.py", line 210, in execute
    branch = super().execute(context)
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/airflow/operators/python.py", line 175, in execute
    return_value = self.execute_callable()
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/airflow/operators/python.py", line 192, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/opt/airflow/dags/foph_dag.py", line 204, in compare
    if not loading.compare(filename, tablename):
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/epigraphhub/data/foph/loading.py", line 68, in compare
    table_date = _table_last_update(table)
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/epigraphhub/data/foph/loading.py", line 123, in _table_last_update
    raise (e)
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/epigraphhub/data/foph/loading.py", line 114, in _table_last_update
    df = pd.read_sql(f"select * from switzerland.foph_{table.lower()}_d;", engine)
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/pandas/io/sql.py", line 592, in read_sql
    return pandas_sql.read_query(
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/pandas/io/sql.py", line 1557, in read_query
    result = self.execute(*args)
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/pandas/io/sql.py", line 1402, in execute
    return self.connectable.execution_options().execute(*args, **kwargs)
  File "<string>", line 2, in execute
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 468, in warned
    return fn(*args, **kwargs)
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3257, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1365, in execute
    return self._exec_driver_sql(
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1669, in _exec_driver_sql
    ret = self._execute_context(
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/opt/conda/envs/epigraphhub/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.InsufficientPrivilege) permission denied for schema switzerland
LINE 1: select * from switzerland.foph_casesvaccpersons_d;
</details>

> PS: btw, in the worst case we could have it as a COPY statement in the dockerfile
I tried to run the sql file directly into the postgres container, all tables and schemas were created, I think only the grants didn't worked.

**Comment by xmnlab** [comment]
> sure, the logs are very similar

I mean the postgresql server log

**Comment by luabida** [comment]
I see, thank you for spotting it out

**Comment by luabida** [comment]
@xmnlab by any chance, do you why is CI accusing `switzerland` schema doesn't exist but the `public.owid_covid` does? https://github.com/thegraphnetwork/EpiGraphHub/actions/runs/4682606854/jobs/8296668536#step:14:4309 I'm getting out of ideas here, the schema creation happens even before the owid_covid table is created and it still says there is no schema `switzerland`..  https://github.com/thegraphnetwork/EpiGraphHub/blob/2d218052494b759d1e4c1dc2a31c579dc8008a95/containers/postgresql/sql/dev/02-epigraphhub.sql#L59

**Comment by xmnlab** [comment]
@luabida as your code is not about anything that would change the potgresql server or the dags, what I can suggest you for now is to move back containers/postgresql/sql/dev/02-epigraphhub.sql and we can merge this PR. and we can take a look into that later

**Comment by luabida** [comment]
@xmnlab selenium tests are passing.. ready to review & merge

**Review by xmnlab — APPROVED** [review]
Changes look good to me  thanks for working on that

**Pull request merged**